# Used Cars on German eBay
- The dataset comes from eBay Kleinanzeigen, a classified section of the German ebay website. It is only a sample of the original dataset consisting of 50,000 data points. 
- Some of the columns include: dateCrawled, name, seller, price, vehicleType, model, kilometer, brand 
- The data will be cleaned and analyzed.

In [2]:
import pandas as pd
import numpy as np
autos = pd.read_csv("autos.csv", encoding='Latin-1')

In [3]:
autos.info()
print(autos.head(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

## Observations
- There are numberous columns with null values.
- The dataset is a mixture of strings and integers. Most of the data points are strings. Some of the columns, such as the odometer and price columns, are strings but would be more useful if changed to integers.
- The name column is messy and would will most likely not be useful in an analysis.
- The column names are not in written in the preferred style. 

In [4]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [5]:
cols = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_of_pictures', 'postal_code',
       'last_seen']
autos.columns = cols
print(autos.head(10))

          date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  pri

### Changes to Column Names
- Column names were changed from camelcase to snakecase to make them easier to work with in Python and for the sake of uniformity.

In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_of_pictures   postal_code  \
count                 50000          50000.0  50000.000000   
unique                   76              NaN           NaN   
top     2016-04-03 00:00:00              NaN           NaN   
freq                   1946              NaN           NaN   
mean                    NaN              0.0  50813.627300   
std                     NaN              0.0  25779.747957   
min                     NaN              0.0   1067.000000   
25%                     NaN              0.0  30451.000000   
50%                     NaN              0.0  49577.000000   
75%                     NaN              0.0  71540.000000   
max                     NaN              0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

## Observations
- The 'seller' and 'offer_type' columns have mostly one value and could probably be dropped. The 'num_of_pictures' column has only one value and can be dropped.
- 'registration_year', 'num_of_pictures', 'postal_code', 'power_ps', and 'registration_month' columns have no unique or top values, so they need to be investigated.
- 'price' and 'odometer' columns are stored as text and must be changed to numeric data.

In [7]:
print(autos['price'].unique())
print(autos['odometer'].unique())

['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']
['150,000km' '70,000km' '50,000km' '80,000km' '10,000km' '30,000km'
 '125,000km' '90,000km' '20,000km' '60,000km' '5,000km' '100,000km'
 '40,000km']


In [8]:
autos['price'] = (autos['price'].str.replace('$','').str.replace(',','').astype(int))
autos['odometer'] = (autos['odometer'].str.replace('km','').str.replace(',','').astype(int))
autos.rename({'price':'price_dollars'}, axis=1, inplace=True)
autos.rename({'odometer':'odometer_km'}, axis=1, inplace=True)

## Price and Odometer Column Cleanup
- I removed the $ symbol and the commas from the 'price' column. Then I converted the values from strings to integers.
- I removed 'km' and the commas from the 'odometer' column. Then I converted the values from strings to integers.

In [10]:
autos['price_dollars'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollars, dtype: float64

In [11]:
price_counts = autos['price_dollars'].value_counts()
print(price_counts.sort_index(ascending=True))

0           1421
1            156
2              3
3              1
5              2
8              1
9              1
10             7
11             2
12             3
13             2
14             1
15             2
17             3
18             1
20             4
25             5
29             1
30             7
35             1
40             6
45             4
47             1
49             4
50            49
55             2
59             1
60             9
65             5
66             1
            ... 
151990         1
155000         1
163500         1
163991         1
169000         1
169999         1
175000         1
180000         1
190000         1
194000         1
197000         1
198000         1
220000         1
250000         1
259000         1
265000         1
295000         1
299000         1
345000         1
350000         1
999990         1
999999         2
1234566        1
1300000        1
3890000        1
10000000       1
11111111       2
12345678      

In [12]:
autos = autos[autos['price_dollars'].between(0,350000)]

## Dropped Price Values
- I dropped prices that were over 350,000 dollars because they seemed extremely high. The dropped price range is  999,990 to 99,999,999.

In [13]:
autos['odometer_km'].describe()

count     49986.000000
mean     125736.506222
std       40038.133399
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [14]:
odometer_counts = autos['odometer_km'].value_counts()
print(odometer_counts.sort_index(ascending=True))

5000        966
10000       264
20000       784
30000       789
40000       818
50000      1025
60000      1164
70000      1230
80000      1436
90000      1757
100000     2168
125000     5169
150000    32416
Name: odometer_km, dtype: int64


## Odometer Reading Obsertations
- All of the odomter values (km) look reasonable, so no rows were removed.

In [15]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49986 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          49986 non-null object
name                  49986 non-null object
seller                49986 non-null object
offer_type            49986 non-null object
price_dollars         49986 non-null int64
ab_test               49986 non-null object
vehicle_type          44894 non-null object
registration_year     49986 non-null int64
gearbox               47310 non-null object
power_ps              49986 non-null int64
model                 47233 non-null object
odometer_km           49986 non-null int64
registration_month    49986 non-null int64
fuel_type             45509 non-null object
brand                 49986 non-null object
unrepaired_damage     40163 non-null object
ad_created            49986 non-null object
num_of_pictures       49986 non-null int64
postal_code           49986 non-null int64
last_seen             49986 non-null object
dtypes: int64(7), 

In [15]:
autos[['date_crawled', 'ad_created', 'last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

## Date Columns
- The 'date_crawled', 'ad_created', and 'last_seen' columns need cleaned a little before any aggregation can be done.
- In order to do so, I will extract only the first 10 values of the string for each date, which will give us the year, month, and day in the following format: yyyy-mm-dd
- Then I will determine the percentage of the ads created on each date.

In [16]:
autos['date_crawled'] = autos['date_crawled'].str[:10]
autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025387
2016-03-06    0.013944
2016-03-07    0.035970
2016-03-08    0.033269
2016-03-09    0.033209
2016-03-10    0.032129
2016-03-11    0.032489
2016-03-12    0.036770
2016-03-13    0.015564
2016-03-14    0.036630
2016-03-15    0.033990
2016-03-16    0.029508
2016-03-17    0.031509
2016-03-18    0.013064
2016-03-19    0.034910
2016-03-20    0.037831
2016-03-21    0.037490
2016-03-22    0.032909
2016-03-23    0.032389
2016-03-24    0.029108
2016-03-25    0.031749
2016-03-26    0.032489
2016-03-27    0.031049
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-30    0.033629
2016-03-31    0.031909
2016-04-01    0.033809
2016-04-02    0.035410
2016-04-03    0.038691
2016-04-04    0.036490
2016-04-05    0.013104
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64

### Observations
- Most of the dates crawled have a similar percentage of listings, around 3-3.5 percent. There are only a few outliers with values around 1-2 percent. 

In [17]:
autos['ad_created'] = autos['ad_created'].str[:10]
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
2015-12-30    0.000020
2016-01-03    0.000020
2016-01-07    0.000020
2016-01-10    0.000040
2016-01-13    0.000020
2016-01-14    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-27    0.000060
2016-01-29    0.000020
2016-02-01    0.000020
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-07    0.000020
2016-02-08    0.000020
2016-02-09    0.000040
2016-02-11    0.000020
2016-02-12    0.000060
2016-02-14    0.000040
2016-02-16    0.000020
2016-02-17    0.000020
2016-02-18    0.000040
2016-02-19    0.000060
2016-02-20    0.000040
2016-02-21    0.000060
                ...   
2016-03-09    0.033229
2016-03-10    0.031869
2016-03-11    0.032789
2016-03-12    0.036610
2016-03-13    0.016925
2016-03-14    0.035230
2016-03-15    0.033749
2016-03-16    0.030008
2016-03-17    0.031189
2016-03-18    0.013724
2016-03-19    0.033849
2016-03-20    0.037871
2016-03-21 

### Observations
- There were, comparatively, very few adds created in 2015 or early 2016. In March of 2016, the percentage of adds jumps from around 0.006 percent to 3 percent. The percentage of adds stays around 3 percent until early April.

In [18]:
autos['last_seen'] = autos['last_seen'].str[:10]
autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-08    0.007582
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-12    0.023807
2016-03-13    0.008983
2016-03-14    0.012804
2016-03-15    0.015884
2016-03-16    0.016445
2016-03-17    0.027928
2016-03-18    0.007422
2016-03-19    0.015744
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-22    0.021586
2016-03-23    0.018585
2016-03-24    0.019565
2016-03-25    0.019205
2016-03-26    0.016965
2016-03-27    0.016024
2016-03-28    0.020846
2016-03-29    0.022326
2016-03-30    0.024847
2016-03-31    0.023827
2016-04-01    0.023106
2016-04-02    0.024887
2016-04-03    0.025367
2016-04-04    0.024627
2016-04-05    0.124275
2016-04-06    0.220982
2016-04-07    0.130957
Name: last_seen, dtype: float64

### Observations
- At the beginning of March the percentages of the last_seen values are relatively low, from 0.1 to 0.9 percent. For the rest of March and into the beginning of April the percentages gover around 1-2.5 percent. 

In [19]:
autos['registration_year'].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

## Observations
- The 'registration_year' column contains some values that do not make sense. The minimum year is 1000, and the maximum value is 9999. So, some rows need to be removed.
- Also, we need to check whether there are any registration years that are after the listing year, since this wouldn't make sense.
- Any registration years outside the range 1900-2016 should be removed. Cars were not mass produced before the early 1900's, and it's very unlikely that someone would be selling a car that was owned before that time. Any dates after 2016 wouldn't make sense since a car cannot be registered before it appears on eBay and is sold. 

In [20]:
autos = autos[autos['registration_year'].between(1900,2016)]
autos['registration_year'].value_counts(normalize=True)

2000    0.069852
2005    0.062792
1999    0.062438
2004    0.057002
2003    0.056794
2006    0.056377
2001    0.056273
2002    0.052753
1998    0.051087
2007    0.047984
2008    0.046464
2009    0.043673
1997    0.042236
2011    0.034030
2010    0.033260
1996    0.030073
2012    0.027553
2016    0.027408
1995    0.027324
2013    0.016786
2014    0.013850
1994    0.013745
1993    0.009268
2015    0.008310
1990    0.008226
1992    0.008122
1991    0.007414
1989    0.003770
1988    0.002957
1985    0.002166
          ...   
1966    0.000458
1977    0.000458
1975    0.000396
1969    0.000396
1965    0.000354
1964    0.000250
1910    0.000187
1963    0.000187
1959    0.000146
1961    0.000125
1956    0.000104
1958    0.000083
1937    0.000083
1962    0.000083
1950    0.000062
1954    0.000042
1941    0.000042
1934    0.000042
1957    0.000042
1951    0.000042
1955    0.000042
1931    0.000021
1953    0.000021
1943    0.000021
1938    0.000021
1939    0.000021
1927    0.000021
1929    0.0000

In [21]:
autos['registration_year'].describe()

count    48016.000000
mean      2002.806002
std          7.306212
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

## Observations about Registration Year
- The lowest percentage for registration year is 1952, followed by 1948, 1926, 1927, 1939 ascending. The years with the highest percentages are 2000 (almost 7 percent), 2005 (6 percent), and 1999 (6 percent). 
- The highest percentages of registration years come from the 1990's and from the 2000's, which is not surprising. 
- Relatively speaking, very few of the cars listed were registered before the mid 1900's. 
- The minimum registration year is 1910 and the max is 2016. 

In [22]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.212117
bmw               0.110026
opel              0.108172
mercedes_benz     0.095364
audi              0.086409
ford              0.069768
renault           0.047359
peugeot           0.029532
fiat              0.025866
seat              0.018181
skoda             0.016036
mazda             0.015141
nissan            0.015099
smart             0.013912
citroen           0.013912
toyota            0.012475
sonstige_autos    0.010892
hyundai           0.009851
volvo             0.009247
mini              0.008643
mitsubishi        0.008143
honda             0.007852
kia               0.007102
alfa_romeo        0.006623
porsche           0.006102
suzuki            0.005915
chevrolet         0.005706
chrysler          0.003665
dacia             0.002562
daihatsu          0.002562
jeep              0.002249
subaru            0.002187
land_rover        0.002041
saab              0.001604
jaguar            0.001583
trabant           0.001562
daewoo            0.001500
r

In [23]:
autos['brand'].describe()

count          48016
unique            40
top       volkswagen
freq           10185
Name: brand, dtype: object

### I will aggregate on all of the brands that make 1% or more of the ads. This includes 17 of the 40 brands.

In [24]:
unique_brands = ['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'smart',
       'citroen', 'toyota', 'sonstige_autos']
brand_mean_prices = {}
for i in unique_brands:
    mean_price = autos['price_dollars'][autos['brand'] == i].mean()
    brand_mean_prices[i] = mean_price
print(brand_mean_prices)

{'nissan': 4664.891034482758, 'sonstige_autos': 10805.078393881453, 'mercedes_benz': 8485.239571958942, 'skoda': 6334.91948051948, 'fiat': 2711.8011272141707, 'audi': 9093.65003615329, 'mazda': 4010.7716643741405, 'bmw': 8102.536248343744, 'citroen': 3699.935628742515, 'toyota': 5115.33388981636, 'volkswagen': 5231.081983308787, 'ford': 3652.095223880597, 'opel': 2876.716403542549, 'renault': 2395.4164467897976, 'seat': 4296.492554410081, 'smart': 3542.706586826347, 'peugeot': 3039.4682651622}


## Observations about Mean Brand Price
- The average prices range from about 2,300 dollars to around 10,000 dollars.
- Most Expensive Brands: Sonstige (m=10805), Audi (m=9030), BMW (m=8102), Mercedes (m=8485), and Skoda (m=6334).
- Least Expensive Brands: Renault (m=2395), Fiat (m=2711), Opel (m=2876), Peugeot (m=3039), Smart (m=3542), Ford (m=3652), and Citroen (m=3699).
- Mid-Range Brands: Toyota (m=5115), Nissan (m=4664), Volkswagen (m=5231), Seat (m=4296), and Mazda (m=4010).

## The Plan
- We want to calculate the mean mileage for each brand and then see if there is any discernible connection between mileage and mean price. 

In [25]:
unique_brands = ['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'smart',
       'citroen', 'toyota', 'sonstige_autos']
brand_mean_mileage = {}
for i in unique_brands:
    mean_mileage = autos['odometer_km'][autos['brand'] == i].mean()
    brand_mean_mileage[i] = mean_mileage
print(brand_mean_mileage)

{'nissan': 118572.41379310345, 'sonstige_autos': 87466.53919694072, 'mercedes_benz': 130856.0821139987, 'skoda': 110954.54545454546, 'fiat': 116553.94524959743, 'audi': 129287.78018799711, 'mazda': 124745.5295735901, 'bmw': 132431.38368351315, 'citroen': 119580.8383233533, 'toyota': 115709.51585976628, 'volkswagen': 128724.10407461954, 'ford': 124068.65671641791, 'opel': 129223.14208702349, 'renault': 128183.81706244503, 'seat': 121563.57388316152, 'smart': 99595.80838323354, 'peugeot': 127136.81241184767}


In [40]:
bmp_series = pd.Series(brand_mean_prices)
bmm_series = pd.Series(brand_mean_mileage)

df = pd.DataFrame(bmp_series, columns=['mean_prices'])
df['mean_mileage'] = bmm_series
df.sort_values('mean_prices')

mean_prices   mean_mileage
renault          2395.416447  128183.817062
fiat             2711.801127  116553.945250
opel             2876.716404  129223.142087
peugeot          3039.468265  127136.812412
smart            3542.706587   99595.808383
ford             3652.095224  124068.656716
citroen          3699.935629  119580.838323
mazda            4010.771664  124745.529574
seat             4296.492554  121563.573883
nissan           4664.891034  118572.413793
toyota           5115.333890  115709.515860
volkswagen       5231.081983  128724.104075
skoda            6334.919481  110954.545455
bmw              8102.536248  132431.383684
mercedes_benz    8485.239572  130856.082114
audi             9093.650036  129287.780188
sonstige_autos  10805.078394   87466.539197

## Observations about Mean Price and Mean Mileage by Top Brands
- We might expect to see a pattern where higher mileage cars are priced lower on average, whereas lower mileage cars are priced higher. For the most part, we do not observe this pattern, at least at this level of analysis.
- Some of the most expensive brands (Audi, Mercedes, BMW) have some of the highest milages (around 130,000 km). But this is not entirely surprising, since we might expect more expensive cars to be more reliable and to hold their value better at higher mileages.
- One interesting outlier is Sonstige, which is the most expensive brand, but where the mileage is, on average, the lowest (87,466 miles). 
- Most of the mid-range cars have average mileages that range from about 115,000 to about 124,000. The exception is Volkswagen, which has an average mileage of 128,724.  
- The lower-end cars are varied and range from around 99,500 to 129,000.